devmapper: Log fstype and mount options during mount error

rhvgoyal opened the pull request:
Right now we only log source and destination (and demsg) if mount operation
fails. fstype and mount options are available easily. It probably is a good
idea to log these as well. Especially sometimes failures can happen due to
mount options.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
I have heard some reports of error saying mount failed and only error message seems to be "invalid argument". I am also curious to know what mount options were used for mount calls. That's why enhancing the error message to output more debug information.

rhvgoyal commented:
https://github.com/openshift/origin/issues/17330

runcom commented:
:+1:

thaJeztah commented:
Note that this will be less of a problem with https://github.com/moby/moby/pull/35514 merged (but could still be useful for existing installs). Also see https://github.com/moby/moby/pull/35527, which may be of interest as well :innocent:

thaJeztah commented:
Nevermind, I was sleeping, this is devicemapper (got my head thinking due to interpreting `fstype` as `d_type` / `fs_type=1`)

thaJeztah commented:
Are there specific filesystems that cause issues? If so, are there filesystems that should be excluded, similar to https://github.com/moby/moby/blob/0a4e793a3da9ba6d20bccfb83f7c48e20a76d895/daemon/graphdriver/overlay/overlay.go#L140-L144 ?

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
/cc @kolyshkin

rhvgoyal commented:
@thaJeztah By the time we hit this error, we are only dealing with supported filesystems only. IOW, this error we will not hit for unsupported filesystems.

kolyshkin commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1201,7 +1201,7 @@ func (devices *DeviceSet) growFS(info *devInfo) error {
 
 	if err := mount.Mount(info.DevName(), fsMountPoint, devices.BaseDeviceFilesystem, options); err != nil {
-		return fmt.Errorf("Error mounting '%s' on '%s': %s\n%v", info.DevName(), fsMountPoint, err, string(dmesg.Dmesg(256)))
+		return fmt.Errorf("Error mounting '%s' on '%s'. fstype=%s options=%s: %s\n%v", info.DevName(), fsMountPoint, devices.BaseDeviceFilesystem, options, err, string(dmesg.Dmesg(256)))
```
The punctuation looks strange. Perhaps change it to
```go
return fmt.Errorf("Error mounting '%s' on '%s' (fstype='%s' options='%s'): %s\n%v", info.DevName(), fsMountPoint, devices.BaseDeviceFilesystem, options, err, string(dmesg.Dmesg(256)))
```

kolyshkin commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -2392,7 +2392,7 @@ func (devices *DeviceSet) MountDevice(hash, path, mountLabel string) error {
 
 	if err := mount.Mount(info.DevName(), path, fstype, options); err != nil {
-		return fmt.Errorf("devmapper: Error mounting '%s' on '%s': %s\n%v", info.DevName(), path, err, string(dmesg.Dmesg(256)))
+		return fmt.Errorf("devmapper: Error mounting '%s' on '%s'. fstype=%s options=%s: %s\n%v", info.DevName(), path, fstype, options, err, string(dmesg.Dmesg(256)))
```
ditto

rhvgoyal committed:
devmapper: Log fstype and mount options during mount error

Right now we only log source and destination (and demsg) if mount operation
fails. fstype and mount options are available easily. It probably is a good
idea to log these as well. Especially sometimes failures can happen due to
mount options.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>

rhvgoyal commented:
@kolyshkin Took care of your comments. PTAL.

kolyshkin reviewed (approved):
LGTM

vieux reviewed (approved):
LGTM
